### ISSUE TITLE
Warn when `match` arm's guard clause has side-effects

### ISSUE THREAD

**Issue by J-ZhengLi** [rust-lang/rust-clippy#11854]
### What it does

First, I don't know if clippy can detect if a function has side-effect or not (perhaps can't?). So, this issue is suggesting just warn on function calls that use `mut` variable as parameter.

When a match arm guard has side-effects, the operation could be executed multiple times, like this example: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8c3aa161f1c9566d169aee20d9c33b17).

The the result... could be a little unexpected.

### Advantage

reduce unexpected behavior

### Drawbacks

- Cannot cover all scenarios, side-effects is a wide concept.
- Could lead to FP, where the function takes mut variables but not actually changing it.

### Example

```rust
fn overlimit(val: &mut i32) -> bool {
    println!("fn with side-effect called");

    *val += 1;
    *val > LIMIT
}

match op {
    MyNum::OkNumber(0) | MyNum::OkNumber(_) if overlimit(&mut val) => {
        // do something
    },
    _ => ()
};
```

**Comment by PartiallyTyped** [comment]
With respect to FPs, this could be a pedantic lint, so it'd be okay I guess.

**Comment by y21** [comment]
This can only happen with OR patterns specifically, right? Perhaps limiting it those could help a bit with FPs.

**Comment by J-ZhengLi** [comment]
> This can only happen with OR patterns specifically, right? Perhaps limiting it those could help a bit with FPs.

yes, I think... otherwise it'll stop after finding one match

**Comment by Jarcho** [comment]
This isn't limited to top level `or` patterns. `Some((x, 0) | (0, x))` will cause the same thing to happen.
